pkg/storageos/deploy_test.go

```diff
@@ -815,6 +815,113 @@ func TestDeployNodeAffinity(t *testing.T) {
 	}
 }
 
+func TestDeployTolerations(t *testing.T) {
+	testCases := []struct {
+		name        string
+		tolerations []corev1.Toleration
+		wantError   bool
+	}{
+		{
+			name: "TolerationOpExists without value",
+			tolerations: []corev1.Toleration{
+				{
+					Key:      "foo",
+					Operator: corev1.TolerationOpExists,
+					Effect:   corev1.TaintEffectNoSchedule,
+				},
+			},
+		},
+		{
+			name: "TolerationOpExists with value",
+			tolerations: []corev1.Toleration{
+				{
+					Key:      "foo",
+					Operator: corev1.TolerationOpExists,
+					Value:    "bar",
+					Effect:   corev1.TaintEffectNoSchedule,
+				},
+			},
+			wantError: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			deploy := &Deployment{
+				stos: &api.StorageOSCluster{
+					Spec: api.StorageOSClusterSpec{
+						Tolerations: tc.tolerations,
+					},
+				},
+			}
+			podSpec := &corev1.PodSpec{}
+			err := deploy.addTolerations(podSpec)
+			if !tc.wantError && err != nil {
+				t.Errorf("expected no error but got one: %v", err)
+			}
+			if tc.wantError && err == nil {
+				t.Errorf("expected error but got none")
+			}
+		})
+	}
+
+	stosCluster := &api.StorageOSCluster{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: gvk.GroupVersion().String(),
+			Kind:       gvk.Kind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "teststos",
+			Namespace: "default",
+		},
+		Spec: api.StorageOSClusterSpec{
+			Tolerations: []corev1.Toleration{
+				{
+					Key:      "foo",
+					Operator: corev1.TolerationOpExists,
+					Effect:   corev1.TaintEffectNoSchedule,
+				},
+				{
+					Key:      "foo",
+					Value:    "bar",
+					Operator: corev1.TolerationOpEqual,
+					Effect:   corev1.TaintEffectNoSchedule,
+				},
+			},
+		},
+	}
+
+	c := fake.NewFakeClient()
+	deploy := NewDeployment(c, stosCluster, nil, testScheme, "", false)
+	deploy.Deploy()
```

Comment:
Can we combine the two separate test sections to one?
This `deploy.Deploy()` will internally call `addTolerations()`. Hence, `Deploy()` will return error when the toleration configuration is invalid. Add an error check here.
Let's create the StorageOSCluster object inside a test run, and set the toleration from the test cases, like in the first test section. With that, you can have exists, equal, with value and without value, all covered in one set of test cases.
Only if there's no error and toleration is specified, you compare the toleration of the deployed pod and the cluster spec.

Reply:
I've tried using deploy.Deploy but it throws errors even when using the correct handling. But I will give it another shot.

Reply:
If the error you get is something like this https://github.com/storageos/cluster-operator/blob/master/pkg/controller/storageoscluster/storageoscluster_controller.go#L224-L232 , you can add a retry. While running through the controller, the controller runs the deployment again in order to ensure all the resources have been deployed.
If it's something else and you're not sure what to do with it, please share the error. We can figure out what to do with it.

Reply:
I've tried to implement your suggestions. I'm now back to the error:

`expected no error but got one: no kind is registered for the type v1alpha1.StorageOSCluster in scheme "k8s.io/client-go/kubernetes/scheme/register.go:60"`

I think the problem is that the test does not register the scheme and therefore underlying functions fail. As storageos does not wrap errors within `deploy.Deploy()`  I'm not able to tell which inner function is causing this.